shared/health/src/main/java/io/pravega/shared/health/impl/HealthConfigImpl.java

```diff
@@ -0,0 +1,188 @@
+ */
+package io.pravega.shared.health.impl;
+
+import com.google.common.base.Preconditions;
+import io.pravega.shared.health.ContributorRegistry;
+import io.pravega.shared.health.HealthConfig;
+import io.pravega.shared.health.StatusAggregator;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The {@link HealthConfig} class allows a client to define a (logical) tree-like structure of {@link HealthComponent}
+ * to {@link HealthComponent} relations.
+ */
+@Slf4j
+public class HealthConfigImpl implements HealthConfig {
+
+    private final Map<String, HealthComponent> components;
+
+    private final Set<HealthComponent> roots;
+
+    private final Map<String, Set<String>> relations;
+
+    private HealthConfigImpl(Map<String, HealthComponent> components, Set<HealthComponent> roots, Map<String, Set<String>> relations) {
+        this.components = components;
+        this.roots = roots;
+        this.relations = relations;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private final Map<String, HealthComponent> components = new HashMap<>();
+
+        private final Map<String, Set<String>> relations = new HashMap<>();
+
+        private final Set<HealthComponent> roots = new HashSet<>();
+
+        /**
+         * Defines a new {@link HealthComponent} to be created.
+         * @param name The name of the to-be {@link HealthComponent}.
+         * @param aggregator The aggregation rule to use.
+         *
+         * @return A reference to the *same* object, with updates to its internal data-structures made.
+         */
+        public Builder define(String name, StatusAggregator aggregator) {
+            if (components.containsKey(name)) {
+                log.warn("Overwriting a pre-existing component definition -- aborting.");
+            } else {
+                HealthComponent component = new HealthComponent(name, aggregator, null);
+                roots.add(component);
+                components.put(name, component);
+                relations.put(name, new HashSet<>());
+            }
+            return this;
+        }
+
+        /**
+         * Defines that there should be a parent-child relation between a {@link HealthComponent} called 'child'
+         * and a {@link HealthComponent} called 'parent'.
+         *
+         * @param child The name of the child {@link HealthComponent}.
+         * @param parent The name of the parent {@link HealthComponent}.
+         * @return A reference to the *same* object, with updates to its internal data-structures made.
+         */
+        public Builder relation(String child, String parent) {
+            if (!components.containsKey(child) || !components.containsKey(parent)) {
+                log.warn("At least one of the components in this relation has not been defined -- aborting.");
```

Comment:
and here. Best if you throw an exception. Nobody will really look at logs except when debugging some production issue, at which point log noise will be really, really distracting.